### ISSUE TITLE
Cannot set `leftmargin` via `\SetTemplateKeys{block}{display}`

### ISSUE THREAD

**Issue by mbertucci47** [latex3/tagging-project#755]
I am trying to implement margin setting for keytheorems with tagging enabled and ran into trouble overriding the `block` settings locally with `\SetTemplateKeys`. The following MWE shows the issue:
```tex
\DocumentMetadata
  {
    lang=en-US,
    pdfversion=2.0,
    pdfstandard=ua-2,
    testphase={phase-III,math,table,title,firstaid}
  }
\documentclass{article} 

\usepackage{kantlipsum}

\SetTemplateKeys{block}{display}{
  leftmargin=2cm, % does nothing
  rightmargin=2cm, % this works
  }

\begin{document}

\kant[1][1-2]
\kant[2][1-2]
\begin{enumerate}%[leftmargin=2cm] % this works
\item abc
\item \kant[1][1-2]
\kant[2][1-2]
\end{enumerate}
    
\end{document}
```
If `leftmargin` or `rightmargin` is set in the optional argument of `enumerate` then it works as expected. The same is true for `rightmargin` if set with `\SetTemplateKeys{block}{display}`. But `leftmargin` set in `\SetTemplateKeys` has no effect. It doesn't matter if this is executed in the preamble or right before the environment.

Am I trying to use this the wrong way? Or is this a bug in the block code? Even in the first case, it's odd that `rightmargin` does something while `leftmargin` does not.

**Comment by u-fischer** [comment]
The block still calls the legacy commands, in this case `\@listi` and that contains
`\leftmargin\leftmargini`.

**Comment by mbertucci47** [comment]
@u-fischer I see, thanks. I understand these interfaces are a work in progress. So is there any way to change the settings besides either passing the key directly to `enumerate` or redefining `\@listi`? I guess setting `\leftmargini` but that assumes I know the current listdepth is 1.

**Comment by FrankMittelbach** [comment]
I think we have conflicting goals here. I think that the block implementation has to make use of the legacy list environments because they are all over the place in existing document classes and as long as those aren't updated (and many will not) this means that either we keep it (which allows for tagging even of those classes) or drop it (which means we rule out those as being usable going forward) without a full replacement of their code.

So my tendency is to just document this as a restriction.

**Comment by mbertucci47** [comment]
@FrankMittelbach I'm a little confused. Isn't the point of separating the `block` and `list` templates to remove any connection between them? At least for kernel commands where you have full control, shouldn't there be a `trueblock` or something that uses its own parameters completely separate from the list parameters?

**Comment by FrankMittelbach** [comment]
> @FrankMittelbach I'm a little confused. Isn't the point of separating the `block` and `list` templates to remove any connection between them? At least for kernel commands where you have full control, shouldn't there be a `trueblock` or something that uses its own parameters completely separate from the list parameters?

@mbertucci47 yes the idea is to separate both and for new classes that do not also implement `\@listi` and friends that would be the case (and the intention is to offer such classes eventually).  However, the goal also is to make existing classes work evne if they do not get updated and that means honoring their `\@listi` settings somehow.  That means you can't overwrite their settings as cleanly is you could otherwise but if we don't do that then their settings would be ignored which means tagging or no tagging would result in possibly severe layout differences.

Some of that behavior might eventually be controlled with something like  `compatibility={...}` in `\DocumentMetadata` but we are not quite there yet.

**Comment by senthilmm** [comment]
I am not sure whether this is a viable solution.

`\DocumentMetadata
  {
    lang=en-US,
    pdfversion=2.0,
    pdfstandard=ua-2,
    testphase={phase-III,math,table,title,firstaid}
  }
\documentclass{article} 

\usepackage{kantlipsum}

%\SetTemplateKeys{block}{display}{
%  leftmargin=2cm, % does nothing
%  rightmargin=2cm, % this works
%  }

\DeclareInstance{block}{list-1}{display}{leftmargin=2cm}

\begin{document}

\kant[1][1-2]
\kant[2][1-2]

\begin{enumerate}%[leftmargin=2cm] % this works
\item abc
\item \kant[1][1-2]
\kant[2][1-2]
\end{enumerate}
    
\end{document}`

**Comment by mbertucci47** [comment]
@senthilmm Yes that does work, thanks. @FrankMittelbach is it supported behavior to call something like `\DeclareInstance{block}{list-1}{display}{leftmargin=2cm}` locally to change the behavior of a single environment? Something essentially equivalent to
```tex
\begingroup
\DeclareInstance{block}{list-1}{display}{leftmargin=2cm}
\begin{enumerate}
\item abc
\item \kant[1][1-2]
\kant[2][1-2]
\end{enumerate}
\endgroup
```

**Comment by FrankMittelbach** [comment]
It works but better would be to use `\EditInstance` in this case. 

@mbertucci47 your use of `\SetTemplateKeys` is simply wrong and not supported on that point, I didn't realized this initially and looking at the documentation it is clear why you thought it is the right thing to dso. The docu is simply wrong and insufficient (but perhaps we put in some checks to make sure that one gets an error rather than strange behavior). 
`\SetTemplateKeys`overwrite value in the key structure but since execution of a template/instance also alters the  key structure, you get this puzzling behavor: a key that is not explicitly set by the template (either as an explicit value or as a default value) can be altered on top-level, but keys that are set by the template can not (or rather can but wil then be overwritten again).

The real and basically perhaps the only purpose of that command is inside a template **after** the keys have been assigned, e.g., if an environment offers an optional argument to specify keys that argument is then passed to the `\SetTemplateKeys` to overwrite what the template normally does, hence `\begin{enumerate}[leftmargin=2cm]` works.
This is why I said the documentation is wrong  and misleading (besides the fact that the example wouldn't work).

**Comment by mbertucci47** [comment]
@FrankMittelbach Thanks for the explanation, it clarified a lot. I have one followup question. If `rightmargin` is also changed for `list-1`, it seems to also affect nested lists. Compare the following examples.
```tex
\DocumentMetadata
  {
    lang=en-US,
    pdfversion=2.0,
    pdfstandard=ua-2,
    testphase={phase-III,math,table,title,firstaid}
  }
\documentclass{article}

\usepackage{kantlipsum}

\begin{document}

\begin{enumerate}
\item \kant[1][1-2]
\begin{enumerate}
\item \kant[1][1-2]
\end{enumerate}
\end{enumerate}

\begingroup
\EditInstance{block}{list-1}{leftmargin=2cm,rightmargin=1cm}
\begin{enumerate}
\item \kant[1][1-2]
\begin{enumerate}
\item \kant[1][1-2]
\end{enumerate}
\end{enumerate}
\endgroup

\begingroup
\EditInstance{block}{list-1}{leftmargin=2cm}
\begin{enumerate}
\item \kant[1][1-2]
\begin{enumerate}
\item \kant[1][1-2]
\end{enumerate}
\end{enumerate}
\endgroup

\end{document}
```
If nothing is specified for `rightmargin` then the nested list just adapts to the outer list right margin. But if `rightmargin` is specified, the nested list adds that value to its already adjusted right margin. Is this intended or is the `rightmargin` value leaking into the nested list incorrectly? I would have thought this is controlled by `list-2` which hasn't been changed. And the same effect is not observed with `leftmargin`.

**Comment by FrankMittelbach** [comment]
> If nothing is specified for rightmargin then the nested list just adapts to the outer list right margin. But if rightmargin is specified, the nested list adds that value to its already adjusted right margin. Is this intended or is the rightmargin value leaking into the nested list incorrectly? I would have thought this is controlled by list-2 which hasn't been changed. And the same effect is not observed with leftmargin.

This is because of the dependencies with the legacy list setting. What happens with lists is the following:
When entering a new nesting  level
- the corresponding `\@list<level>`, e.g., `\@listii`is executed
- then the instance values overwrite
- then any local optional argument settings overwrite

Now the defaults in `size10.clo`and the like have
```
\def\@listii {\leftmargin\leftmarginii
              \labelwidth\leftmarginii
              \advance\labelwidth-\labelsep
              \topsep    4\p@ \@plus2\p@ \@minus\p@
              \parsep    2\p@ \@plus\p@  \@minus\p@
              \itemsep   \parsep}
```
So leftmargin is set to `\leftmarginii` so your value of 2cm isn't any longer active.
However `\rightmargin` is not  touched by the settings and `list-2` sets  `rightmargin=\rightmargin` (which means inherit from previous level) so it now remains 1cm-

So to prevent that you would need to specify `rightmargin=0pt` on `list-2` as well. 

Same would happen in standard 2e right now if one adds a value into `\@listi` without also adjusting `\@listii` etc.

This certainly needs a good documentation and going forward I hope we will see classes not using `\@listi`but just the instance setups and then things become more sane for changing. Another adjustment that I'm contemplating is a key that allows you to supress the use of `\@listi` and friends so that all values only depend on what is specified on the instances.

**Comment by mbertucci47** [comment]
@FrankMittelbach Again, thanks for the detailed explanation. To me that procedure is counterintuitive; I wonder why lists were set up that way (I guess by Leslie Lamport?). I was especially surprised because as I mentioned, I'm trying to implement margin-changing for tagged theorems and the problem is still there when you replace `list-1` with `theoremblock-0` and have a list inside a theorem. Of course, you already said the new theorem code still uses the list parameters so I should have known better.

**Comment by FrankMittelbach** [comment]
@mbertucci47 can you explain in a bit more detail what you try to achieve? Eg with an example of intended output?
This clearly goes beyond the initial issue, butt we might well continue here I think.

**Comment by mbertucci47** [comment]
@FrankMittelbach Sorry I must have missed your last comment. Here is an example showing what I'm trying to achieve. The first two are normal behavior of lists inside a theorem, what I'd expect. In the third and fourth, calling
```tex
\EditInstance{block}{theoremblock-0}
  {
    leftmargin=2cm,
    rightmargin=2cm,
  }
```
causes the right margin of lists inside the theorem do be increasingly indented, which is not what I want. Since theorems are conceptually not lists, it would be best if they did not use the list settings. This can be "fixed" by simultaneously setting
```tex
\EditInstance{block}{list-1}{rightmargin=0pt}
```
but if for whatever reason the user or class had set a different value for `rightmargin`, this would incorrectly override it.
```tex
% !TeX program = txs:///lualatex-dev
\DocumentMetadata
  {
    lang=en-US,
    pdfversion=2.0,
    pdfstandard=ua-2,
    testphase={phase-III,math,table,title,firstaid}
  }
\documentclass{article}

\usepackage{amsthm}
\usepackage{kantlipsum}

\newtheorem{theorem}{Theorem}

\begin{document}

% normal behavior
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}

% normal behavior
\begin{theorem}
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}
\end{theorem}

\clearpage

% behavior I don't want
\begingroup
\EditInstance{block}{theoremblock-0}
  {
    leftmargin=2cm,
    rightmargin=2cm,
  }
\begin{theorem}
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}
\end{theorem}
\endgroup

% behavior I want
\begingroup
\EditInstance{block}{theoremblock-0}
  {
    leftmargin=2cm,
    rightmargin=2cm,
  }
\EditInstance{block}{list-1}{rightmargin=0pt}
\begin{theorem}
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}
\end{theorem}
\endgroup

\end{document}
```

### First page
<img width="676" alt="Image" src="https://github.com/user-attachments/assets/69b1e7d9-e7d3-4fcc-ae9f-7f04b96d2c3a" />

### Second page
<img width="409" alt="Image" src="https://github.com/user-attachments/assets/1661591c-68ec-4170-9938-deb3fb81cfce" />

**Comment by mbertucci47** [comment]
As far as I can see the issue is still there with the 2025-11-01 format. Is this a "won't fix" for historical reasons? I still think it is very counterintuitive that a single top-level setting of `right-margin=2cm` for the theoremblock would affect nested lists in this way.

Here is the same code with the key and template names updated for the new latex-lab-block.
```tex
\DocumentMetadata
  {
    lang=en-US,
    pdfversion=2.0,
    pdfstandard=ua-2,
    tagging=on,
  }
\documentclass{article}

\usepackage{amsthm}
\usepackage{kantlipsum}

\newtheorem{theorem}{Theorem}

\begin{document}

% normal behavior
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}

% normal behavior
\begin{theorem}
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}
\end{theorem}

\clearpage

% behavior I don't want
\begingroup
\EditInstance{block}{theoremblock-0}
  {
    left-margin=2cm,
    right-margin=2cm,
  }
\begin{theorem}
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}
\end{theorem}
\endgroup

% behavior I want
\begingroup
\EditInstance{block}{theoremblock-0}
  {
    left-margin=2cm,
    right-margin=2cm,
  }
\EditInstance{block}{listblock-1}{right-margin=0pt}
\begin{theorem}
\kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\begin{enumerate}
\item \kant[1][1]
\end{enumerate}
\end{enumerate}
\end{theorem}
\endgroup

\end{document}
```
